WordPress/gutenberg
# SelectControl: Infer `value` type from `options`

Updated take on #60293 (props @mikeybinns)

## What?

Tighten the TypeScript type checking on `SelectControl`, so the `value` type matches in the `value`, `options`, and `onChange` props.

See Dev Notes at the bottom for how this looks like to a consumer.

## Why?

It can catch potential bugs.

## How?

Adding a generic that can either be inferred from the `options` array (if it is immutable), or be set explicitly.

## Testing Instructions

See some of the static type tests in the unit test file, and play around by adding your own cases and see what errors.

---

## ✍️ Dev Note

### Stricter type checking in `SelectControl`

In TypeScript codebases, `SelectControl` will now be able to type check whether the `value` type matches in the `value`, `options`, and `onChange` props.

<details>
<summary>This will be inferred automatically if the <code>options</code> array is declared as immutable.</summary>
<p>

```tsx
<SelectControl
	// @ts-expect-error "baz" is not in `options` array
	value="baz"
	options={ [ // array literal that cannot be mutated
		{
			value: 'foo',
			label: 'Foo',
		},
		{
			value: 'bar',
			label: 'Bar',
		},
	] }
/>;
```

```tsx
const options = [
	{
		value: 'foo',
		label: 'Foo',
	},
	{
		value: 'bar',
		label: 'Bar',
	},
] as const; // const assertion to mark as readonly

<SelectControl
	// @ts-expect-error "baz" is not in `options` array
	value="baz"
	options={ options }
/>;
```

</p>
</details>

<details>
<summary>If the <code>options</code> array is mutable, <code>value</code> will remain a simple <code>string</code> type.</summary>
<p>

This is effectively the same type checking behavior as before.

```tsx
const options = [
	{
		value: 'foo',
		label: 'Foo',
	},
	{
		value: 'bar',
		label: 'Bar',
	},
];

<SelectControl
	// No error, because `value` has type `string`
	value="baz"
	options={ options }
/>;
```

</p>
</details>

<details>
<summary>You can also set an explicit <code>value</code> type.</summary>
<p>

```tsx
<SelectControl< 'foo' | 'bar' >
	// @ts-expect-error "baz" is not in `'foo' | 'bar'`
	value="baz"
	// @ts-expect-error "baz" is not in `'foo' | 'bar'`
	options={ [ { value: 'baz', label: 'Baz' } ] }
/>;
```

</p>
</details>
